validation/object_test.go

```diff
@@ -0,0 +1,574 @@
+			},
+			expectError:           true,
+			expectValidationError: true,
+		},
+		{
+			name:                     "dry run validation fails",
+			allowNamespaceEscalation: true,
+			obj: &unstructured.Unstructured{
+				Object: map[string]interface{}{
+					"apiVersion": "v1",
+					"kind":       "ConfigMap",
+					"metadata": map[string]interface{}{
+						"name":      "test",
+						"namespace": "default",
+					},
+				},
+			},
+			owner: &unstructured.Unstructured{
+				Object: map[string]interface{}{
+					"metadata": map[string]interface{}{
+						"namespace": "default",
+					},
+				},
+			},
+			mockSetup: func(_ *mockRestMapper, writer *mockWriter) {
+				statusErr := &apimachineryerrors.StatusError{
+					ErrStatus: metav1.Status{
+						Reason: metav1.StatusReasonInvalid,
+					},
+				}
+				writer.On("Patch", mock.Anything, mock.Anything, mock.Anything, mock.Anything).Return(statusErr)
+			},
+			expectError:           true,
+			expectValidationError: true,
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			t.Parallel()
+
+			restMapper := &mockRestMapper{}
+			writer := &mockWriter{}
+			test.mockSetup(restMapper, writer)
+
+			validator := &ObjectValidator{
+				restMapper:               restMapper,
+				writer:                   writer,
+				allowNamespaceEscalation: test.allowNamespaceEscalation,
+			}
+
+			err := validator.Validate(t.Context(), test.owner, test.obj)
+
+			if test.expectError {
+				require.Error(t, err)
+
+				if test.expectValidationError {
+					var objErr *ObjectValidationError
+
+					require.ErrorAs(t, err, &objErr)
+				}
+			} else {
+				require.NoError(t, err)
+			}
+
+			restMapper.AssertExpectations(t)
+			writer.AssertExpectations(t)
+		})
+	}
+}
+
+func TestMustBeNamespaceScopedResourceError(t *testing.T) {
+	t.Parallel()
+
+	err := MustBeNamespaceScopedResourceError{}
+	assert.Equal(t, "object must be namespace-scoped", err.Error())
+}
+
+func TestMustBeInNamespaceError(t *testing.T) {
+	t.Parallel()
+
+	err := MustBeInNamespaceError{
+		ExpectedNamespace: "default",
+		ActualNamespace:   "other",
+	}
+	assert.Equal(t, "object must be in namespace \"default\", actual \"other\"", err.Error())
+}
+
+func TestValidateNamespace(t *testing.T) {
+	t.Parallel()
+
+	tests := []struct {
+		name          string
+		namespace     string
+		obj           *unstructured.Unstructured
+		mockSetup     func(*mockRestMapper)
+		expectedError error
+		expectNoError bool
+	}{
+		{
+			name:      "no namespace limitation",
+			namespace: "",
+			obj: &unstructured.Unstructured{
+				Object: map[string]interface{}{
+					"apiVersion": "v1",
+					"kind":       "ConfigMap",
+					"metadata": map[string]interface{}{
+						"name":      "test",
+						"namespace": "any",
+					},
+				},
+			},
+			mockSetup: func(restMapper *mockRestMapper) {
+				// Even with empty namespace, RESTMapping is called first (potential bug in implementation)
```

Comment:
As in: "since there is no namespace limitation, there is no need to figure out the scope of the object's API"?

Reply:
Fixed, don't know why I've not cached that before.